lib/ansible/plugins/connection/saltstack.py

```diff
@@ -64,6 +64,14 @@ def exec_command(self, cmd, sudoable=False, in_data=None):
         if in_data:
             raise errors.AnsibleError("Internal Error: this module does not support optimized module pipelining")
 
+        # Fixes an issue where tilde '~' is not resolved to the actual user home
+        get_current_user_home = self.client.cmd(self.host, 'cmd.exec_code_all', ['bash', 'echo ~'])[self.host]
```

Comment:
i don't know enough about salt, but wouldn't the current user have `os.environ['HOME'] ` as a 'simple option' to find it?

Reply:
I might use `echo $HOME` instead of `echo ~` but i cannot depend on the value of `os.environ['HOME']` through ansible's code if that is what you mean. As ansible and salt might be running with different users. Would you like me to use `echo $HOME` instead ?

Reply:
i'm unsure of 'which user' you are looking for, but if it is the 'current one being executed', the environ should 'just work'tm. $HOME and ~ are supposed to be equal, sadly some systems don't work with one or the other, i.e cloud-init/solaris/etc ... so I cannot really say which one is best, it depends on the target. 

This is why it is configurable in ansible, it defaults to `~` but allows users for those system to set it as `$HOME`.

Reply:
i am trying to get the user's home that is being used by saltstack not ansible.
So you might be using ansible as root user but your minion is running with a non root user so in this case the home user will be different. so i though it could be better if `echo ~` evaluated on salt's behalf then i could use it's value in other operations.

Reply:
FYI `~` does not always work (i.e cloud init environments and solaris fail at it), that is why we normally use `$HOME`

you are making many assumptions that might work most of the time (i.e bash) but these are not guaranteed, normally we keep code like this in a 'shell plugin' .. which might be needed here, since we are really executing in 'salt minion' vs actual shell